packages/foam-vscode/src/core/janitor/convert-links-format.ts

```diff
@@ -0,0 +1,103 @@
+  range: Range /* old range */;
+}
+
+/**
+ * convert a link based on its workspace and the note containing it.
+ * According to targetFormat parameter to decide output format. If link.type === targetFormat, then it simply copy
+ * the rawText into LinkReplace. Therefore, it's recommanded to filter before conversion.
+ * If targetFormat isn't supported, or the target resource pointed by link cannot be found, the function will throw
+ * exception.
+ * @param link
+ * @param targetFormat
+ * @param workspace
+ * @param note
+ * @returns LinkReplace { newText: string; range: Range; }
+ */
+export function convertLinkFormat(
+  link: ResourceLink,
+  targetFormat: string,
+  workspace: FoamWorkspace,
+  note: Resource | URI
+): LinkReplace {
+  const resource = note instanceof URI ? workspace.find(note) : note;
+  const targetUri = workspace.resolveLink(resource, link);
+  /* If it's already the target format or a placeholder, no transformation happens */
+  if (link.type === targetFormat || targetUri.scheme === 'placeholder') {
+    return {
+      newText: link.rawText,
+      range: link.range,
+    };
+  }
+
+  let { target, section, alias } = MarkdownLink.analyzeLink(link);
+  let sectionDivider = section ? '#' : '';
+  let aliasDivider = alias ? '|' : '';
+  let embed = link.isEmbed ? '!' : '';
+
+  if (isNone(targetUri)) {
+    throw new Error(
+      `Unexpected state: link to: "${link.rawText}" is not resolvable`
+    );
+  }
+
+  const targetRes = workspace.find(targetUri);
+  let relativeUri = targetRes.uri.relativeTo(resource.uri.getDirectory());
+
+  if (targetFormat === 'wikilink') {
+    /* remove extension if possible, then get the basename to prevent from filename conflict */
+    if (relativeUri.path.endsWith(workspace.defaultExtension)) {
+      relativeUri = relativeUri.changeExtension('*', '');
+    }
+    target = relativeUri.getBasename();
+
+    return {
+      newText: `${embed}[[${target}${sectionDivider}${section}${aliasDivider}${alias}]]`,
+      range: link.range,
+    };
+  }
+
+  if (targetFormat === 'link') {
+    /* if alias is empty, construct one as target#section */
+    if (alias === '') {
+      /* in page anchor have no filename */
+      if (relativeUri.getBasename() === resource.uri.getBasename()) {
+        target = '';
+      }
+      alias = `${target}${sectionDivider}${section}`;
+    }
+
+    /* construct url */
+    let url = relativeUri.path;
+    /* in page anchor have no filename */
+    if (relativeUri.getBasename() === resource.uri.getBasename()) {
+      url = '';
+    }
+    if (sectionDivider === '#') {
+      url = `${url}${sectionDivider}${section}`;
+    }
+    if (url.indexOf(' ') > 0) {
+      url = `<${url}>`;
+    }
+
+    /* if it's originally an embeded note, the markdown link shouldn't be embeded */
+    if (embed && targetRes.type === 'note') {
+      embed = '';
+    }
+    return {
+      newText: `${embed}[${alias}](${url})`,
+      range: link.range,
+    };
+  }
```

Comment:
I haven't tested it locally, but I wonder if you could be able to replace this whole section with
```
const newLink = {...link, type: targetType}
const edit = MarkdownLink.createUpdateLinkEdit(newLink)
```

Reply:
I think it's impossible. The first line of `MarkdownLink.createUpdateLinkEdit` is

```typescript
const { target, section, alias } = MarkdownLink.analyzeLink(link);
```

which extracts three components via regex from link.rawText based on link.type. If the pattern doesn't match, three of them may not be extracted correctly.


Moreover, when dealing with conversion from wikilink to link, we have to retrieve the relative path to target, which is usually a basename in wikilink's rawText. Therefore, we must have `workspace` and `note: Resource | URI`.

Reply:
Yeah, good point. Basically instead in`createUpdateLinkEdit` we need to add `type?: 'wikilink' | 'link'` to the `delta` parameter, and take it from there. Does that sound good?